File: src/tools/ask_user.ts
import type { ToolDef, ToolResult, ToolExecutionContext } from "./registry.js";
import { ToolDeniedError } from "./registry.js";
import { register } from "./registry.js";

interface AskUserArgs {
  question: string;
  options: Array<{ label: string; description: string }>;
  multiSelect?: boolean;
}

export const askUserTool: ToolDef<AskUserArgs> = {
  name: "AskUser",
  description:
    "Ask the user a question with predefined options. Use this when you need the user to choose from a set of alternatives — for example, selecting a library, choosing an approach, or clarifying requirements. Set multiSelect to true when the user may select multiple options. The user can select options or reject all options and type their own response.",
  requiresPermission: false,
  readOnly: true,
  interactive: true,
  input_schema: {
    type: "object" as const,
    properties: {
      question: {
        type: "string",
        description:
          "The question to ask the user. Should be clear and specific.",
      },
      options: {
        type: "array",
        items: {
          type: "object",
          properties: {
            label: {
              type: "string",
              description: "Short label for the option, should be unique",
            },
            description: {
              type: "string",
              description:
                "Explanation of what this option means or what will happen if chosen.",
            },
          },
          required: ["label", "description"],
        },
        minItems: 2,
        maxItems: 4,
        description:
          "2-4 mutually exclusive options for the user to choose from.",
      },
      multiSelect: {
        type: "boolean",
        description:
          "Set to true to allow multiple options to be selected. Default false means single selection.",
      },
    },
    required: ["question", "options"],
  },
  execute: async (
    args: AskUserArgs,
    context?: ToolExecutionContext,
  ): Promise<ToolResult> => {
    const question = args.question;
    const options = args.options;
    if (!Array.isArray(options)) {
      return {
        output: `Error: AskUser tool requires 'options' to be an array, received: ${JSON.stringify(args.options)}`,
      };
    }
    const multiSelect = args.multiSelect ?? false;

    const answer = await context?.prompter?.prompt({
      message: question,
      options: options.map((o) => ({ ...o, value: o.label })),
      multiSelect,
    });

    if (!answer) {
      throw new ToolDeniedError(
        "AskUser",
        question,
        "User cancelled the question",
      );
    }
    return { output: `User selected: "${answer}"` };
  },
};
register(askUserTool);

File: internal/tools/ask_user.go
package tools

import (
	"context"
	"fmt"

	"minicode/internal/domain"
)

// AskUserTool asks the user a question with predefined options.
type AskUserTool struct{}

// NewAskUserTool creates an AskUserTool.
func NewAskUserTool() *AskUserTool {
	return &AskUserTool{}
}

func (t *AskUserTool) Name() string                { return "AskUser" }
func (t *AskUserTool) Description() string         { return "Ask the user a question with predefined options." }
func (t *AskUserTool) RequiresPermission() bool    { return false }
func (t *AskUserTool) Requires() []ToolRequirement { return nil }

func (t *AskUserTool) InputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"question":    map[string]any{"type": "string"},
			"multiSelect": map[string]any{"type": "boolean"},
			"options": map[string]any{
				"type": "array",
				"items": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"label":       map[string]any{"type": "string"},
						"description": map[string]any{"type": "string"},
					},
					"required": []string{"label", "description"},
				},
			},
		},
		"required": []string{"question", "options"},
	}
}

func (t *AskUserTool) Execute(ctx context.Context, args map[string]any, tc ToolContext) (domain.ToolResult, error) {
	question, _ := args["question"].(string)
	multiSelect, _ := args["multiSelect"].(bool)
	rawOptions, _ := args["options"].([]any)

	if tc.AskUserFn == nil {
		tc.LogErr("AskUser: no prompter available")
		return domain.ToolResult{Output: "Error: No user prompter available"}, nil
	}

	var options []domain.AskOption
	for _, opt := range rawOptions {
		if m, ok := opt.(map[string]any); ok {
			label, _ := m["label"].(string)
			desc, _ := m["description"].(string)
			if label != "" {
				options = append(options, domain.AskOption{Label: label, Description: desc})
			}
		}
	}

	if len(options) == 0 {
		tc.LogErr("AskUser: no valid options", "raw_options", len(rawOptions))
		return domain.ToolResult{Output: "Error: No valid options provided"}, nil
	}

	tc.Log("AskUser: presenting question", "question", question, "options", len(options), "multi", multiSelect)

	answer := tc.AskUserFn(question, options, multiSelect)
	if answer == "" {
		tc.Log("AskUser: user cancelled", "question", question)
		return domain.ToolResult{}, &ToolDeniedError{
			ToolName: "AskUser",
			Reason:   "User cancelled the question",
		}
	}

	tc.Log("AskUser: user answered", "question", question, "answer", answer)
	return domain.ToolResult{Output: fmt.Sprintf(`User selected: "%s"`, answer)}, nil
}

func init() { Register(NewAskUserTool()) }
